src/gui/sdlmain.cpp

```diff
@@ -2216,47 +2287,119 @@ void GFX_Events() {
+				 * seems incorrect on KDE and sometimes on MATE
+				 *
+				 * Note that on Windows/Linux-X11/Wayland/macOS,
+				 * event is fired after toggling between full vs
+				 * windowed modes. However this is never fired
+				 * on the Raspberry Pi (when rendering to the
+				 * Framebuffer); therefore we rely on the
+				 * FOCUS_GAINED event to catch window startup
+				 * and size toggles.
+				 */
+				GFX_ResetScreen();
+				GFX_UpdateMouseState();
+				continue;
+
+			case SDL_WINDOWEVENT_FOCUS_GAINED:
+				DEBUG_LOG_MSG("SDL: Window has gained keyboard focus");
+				SetPriority(sdl.priority.focus);
+				CPU_Disable_SkipAutoAdjust();
+				GFX_UpdateMouseState();
+				continue;
+
+			case SDL_WINDOWEVENT_FOCUS_LOST:
+				DEBUG_LOG_MSG("SDL: Window has lost keyboard focus");
 #ifdef WIN32
-					if (sdl.desktop.fullscreen) {
-						VGA_KillDrawing();
-						GFX_ForceFullscreenExit();
-					}
+				if (sdl.desktop.fullscreen) {
+					VGA_KillDrawing();
+					GFX_ForceFullscreenExit();
+				}
 #endif
-					SetPriority(sdl.priority.nofocus);
-					GFX_LosingFocus();
-					CPU_Enable_SkipAutoAdjust();
-					sdl.mouse.has_focus = false;
-					break;
-				default: ;
+				SetPriority(sdl.priority.nofocus);
+				GFX_LosingFocus();
+				CPU_Enable_SkipAutoAdjust();
+				sdl.mouse.has_focus = false;
+				break;
+
+			case SDL_WINDOWEVENT_ENTER:
+				DEBUG_LOG_MSG("SDL: Window has gained mouse focus");
+				continue;
+
+			case SDL_WINDOWEVENT_LEAVE:
+				DEBUG_LOG_MSG("SDL: Window has lost mouse focus");
+				continue;
+
+			case SDL_WINDOWEVENT_SHOWN:
+				DEBUG_LOG_MSG("SDL: Window has been shown");
+				continue;
+
+			case SDL_WINDOWEVENT_HIDDEN:
+				DEBUG_LOG_MSG("SDL: Window has been hidden");
+				continue;
+
+			case SDL_WINDOWEVENT_MOVED:
+				DEBUG_LOG_MSG("SDL: Window has been moved to %d, %d",
+				              event.window.data1,
+				              event.window.data2);
+				continue;
+
+			case SDL_WINDOWEVENT_SIZE_CHANGED:
+				DEBUG_LOG_MSG("SDL: The window size has changed");
+				// The window size has changed either as a
+				// result of an API call or through the system
+				// or user changing the window size.
+				FinalizeWindowState();
+				continue;
+
+			case SDL_WINDOWEVENT_MINIMIZED:
+				DEBUG_LOG_MSG("SDL: Window has been minimized");
+				break;
+
+			case SDL_WINDOWEVENT_MAXIMIZED:
+				DEBUG_LOG_MSG("SDL: Window has been maximized");
+				continue;
+
+			case SDL_WINDOWEVENT_CLOSE:
+				DEBUG_LOG_MSG("SDL: The window manager "
+				              "requests that the window be "
+				              "closed");
+				continue;
+
+#if SDL_VERSION_ATLEAST(2, 0, 5)
+			case SDL_WINDOWEVENT_TAKE_FOCUS:
+				DEBUG_LOG_MSG("SDL: Window is being offered a focus");
+				// should SetWindowInputFocus() on itself or a
+				// subwindow, or ignore
+				continue;
+
+			case SDL_WINDOWEVENT_HIT_TEST:
+				DEBUG_LOG_MSG("SDL: Window had a hit test that "
+				              "wasn't SDL_HITTEST_NORMAL");
+				continue;
+#endif
```

Comment:
Awesome; really like this "API extension" to cover newer SDL versions!

Reply:
Once Ubuntu 16.04 LTS will reach EOL (which will also mean EOL for it's many, many derivative versions such as older Mint releases), I think I'll just update minimum required SDL version to 2.0.5 and remove these ifdefs. Until then - they are annoying, but we can work with it :).

Only 1 more year until 16.04 EOL…